azkaban-common/src/main/java/azkaban/project/DependencyFetcher.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright 2017 LinkedIn Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ */
+
+package azkaban.project;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Throwables;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Properties;
+import javax.xml.bind.annotation.adapters.HexBinaryAdapter;
+import org.apache.log4j.Logger;
+
+import static com.google.common.base.Preconditions.*;
+import static java.util.Objects.*;
+
+
+/**
+ * Helper class to fetch an external dependency and prepare it as per specifications
+ */
+public class DependencyFetcher {
+  private static final Logger log = Logger.getLogger(DependencyFetcher.class);
+  private static final String TMP_FILENAME = "data.tmp";
+
+  private final ProjectManagerConfig config;
+
+  public DependencyFetcher(ProjectManagerConfig config) {
+    this.config = config;
+  }
+
+  @VisibleForTesting
+  void fetchDependency(URI uri) {
+    final Path tmpFilePath = tmpFilePath();
+
+    // Download the resource
+    fetchUri(uri, tmpFilePath);
+
+    // key is the assumed to be unique and will be used to represent this data in Storage API
+    String key = requireNonNull(computeHash(tmpFilePath));
+
+    // Organize content
+    Path target = Paths.get(tmpDir().toString(), key);
+    if (Files.exists(target)) {
+      log.debug(String.format("Already Cached. Skipping [URI: %s, KEY: %s]", uri, key));
```

Comment:
This is gonna be the more common case. I don't want to pollute the logs for this.